packages/smart-contracts/src/contracts/ERC20EscrowToPay.sol

```diff
@@ -0,0 +1,386 @@
+     * @param _paymentRef Reference of the Invoice related.
+     * @dev Uses modifiers OnlyPayer and IsNotFrozen.
+     * @dev unlockDate is set with block.timestamp + twelve months..
+     */
+    function freezeRequest(bytes memory _paymentRef) 
+        external
+        OnlyPayer(_paymentRef)
+        IsInEscrow(_paymentRef)
+        IsNotFrozen(_paymentRef)
+    {
+        if (requestMapping[_paymentRef].emergencyState) 
+        {
+            requestMapping[_paymentRef].emergencyState = false;
+            requestMapping[_paymentRef].emergencyClaimDate = 0;
+            emit RevertedEmergencyClaim(_paymentRef);
+        }
+
+        requestMapping[_paymentRef].isFrozen = true; 
+        requestMapping[_paymentRef].unlockDate = block.timestamp + 52 weeks;
+
+        emit RequestFrozen(_paymentRef);
+    }
+
+    /**
+     * @notice Closes an open escrow and pays the request to payee.
+     * @param _paymentRef Reference of the related Invoice.
+     * @dev Uses OnlyPayer, modifiers IsInEscrow, IsNOtInEmergencyState and IsNotFrozen.
+     */
+    function payRequestFromEscrow(bytes memory _paymentRef) 
+        external 
+        OnlyPayer(_paymentRef) 
+        IsInEscrow(_paymentRef)
+        IsNotInEmergencyState(_paymentRef)
+        IsNotFrozen(_paymentRef) 
+    {
+        require(_withdraw(_paymentRef, requestMapping[_paymentRef].payee), "Withdraw Failed!");
+
+        emit RequestWithdrawnFromEscrow(_paymentRef);  
+    }
+
+    /**
+     * @notice Allows the payee to initiate an emergency claim after a six months lockperiod .
+     * @param _paymentRef Reference of the related Invoice.
+     * @dev Uses modifiers IsInEscrow, IsNotFrozen.
+     */
+    function initiateEmergencyClaim(bytes memory _paymentRef) 
+        external
+        OnlyPayee(_paymentRef)
+        IsInEscrow(_paymentRef)
+        IsNotInEmergencyState(_paymentRef)
+        IsNotFrozen(_paymentRef)
+    {
+        requestMapping[_paymentRef].emergencyClaimDate = block.timestamp + 24 weeks;
+        requestMapping[_paymentRef].emergencyState = true;
+
+        emit InitiatedEmergencyClaim(_paymentRef);
+    }
+
+    /**
+     * @notice Allows the payee claim funds after a six months emergency lockperiod .
+     * @param _paymentRef Reference of the related Invoice.
+     * @dev Uses modifiers IsInEscrow, IsNotFrozen.
+     */
+    function completeEmergencyClaim(bytes memory _paymentRef) 
+        external
+        OnlyPayee(_paymentRef)
+        IsInEscrow(_paymentRef)
+        IsNotFrozen(_paymentRef)
+    {
+        require(requestMapping[_paymentRef].emergencyState && 
+            requestMapping[_paymentRef].emergencyClaimDate <= block.timestamp, 
+            "Not yet!"
+        );
+
+        requestMapping[_paymentRef].emergencyState = false;
+        requestMapping[_paymentRef].emergencyClaimDate = 0;
+
+        require(_withdraw(_paymentRef, requestMapping[_paymentRef].payee), "Withdraw failed!");
+
+        emit EmergencyClaimComplete(_paymentRef);
```

Comment:
I don't think that's necessary because `_withdraw` already emits an event

Reply:
```suggestion
```